mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala

```diff
@@ -211,14 +213,17 @@ private object IDFModel {
     val n = v.size
     v match {
       case SparseVector(size, indices, values) =>
+        val newElements = new ArrayBuffer[(Int, Double)]
         val nnz = indices.size
-        val newValues = new Array[Double](nnz)
         var k = 0
         while (k < nnz) {
-          newValues(k) = values(k) * idf(indices(k))
+          val newValue = values(k) * idf(indices(k))
```

Comment:
As the diff shows, the existing code is already calling `idf(indices(k))`. That call may indeed be expensive and represent a potential optimization, but it's distinct from the problem this PR is intended to address. Seems like there might be room for a second JIRA/PR to handle the issue you identified.